File: Assignment 8/Python/CityDB.py
class BinTree(object):
    def __init__(self):
        self.data = 0
        self.left = None
        self.right = None
        self.isleaf = True
    def setdata(self,data):
        self.data = data
    def setleft(self,left):
        self.left=left
    def setright(self,right):
        self.right=right
    def __str__(self):
        if self.data:
            return str(self.data.citycode)
        return ""
    
    
class CityRecord(object):
    def __init__(self, a,b,c):
        self.citycode = a
        self.x = b
        self.y = c
    def __str__(self):
        return "City Code: %s, X,Y: (%s,%s)" % (self.citycode,self.x,self.y)
    def xyIdentical(self, x,y):
        if self.x == x and self.y == y:
            return True
        return False

class CityDB(object):
    def __init__(self):
        self.rootptr = None

    def search(self,rootptr, item1,item2=None):
        if rootptr == None:
            return False
        if item2: #city coords
            if rootptr.data.xyIdentical(item1,item2):
                return True
        else:
            if rootptr.data.citycode == item1:
                return True
        if self.search(rootptr.left,item1,item2):
            return True
        if self.search(rootptr.right,item1,item2):
            return True
        return False

    
    def insert(self,cityRecord):
        if self.rootptr == None:
            self.rootptr = BinTree()
            self.rootptr.data = cityRecord
        else:
            self.insertRecursive(cityRecord,self.rootptr)
            
    def insertRecursive(self,cityRecord,rootptr):
        if rootptr == 0: #this should never happen.
            raise OSError
        if cityRecord.citycode <= rootptr.data.citycode:
            if rootptr.left == None:
                rootptr.left = BinTree()
                rootptr.left.data = cityRecord
                rootptr.isleaf = False
            else:
                self.insertRecursive(cityRecord,rootptr.left)
        else:
            if rootptr.right == None:
                rootptr.right = BinTree()
                rootptr.right.data = cityRecord
                rootptr.isleaf = False
            else:
                self.insertRecursive(cityRecord,rootptr.right)

    def remove(rootptr, city):
        if rootptr.data == city:
            
        pass
    def remove_max(

    def __str__(self):
        self.printData(self.rootptr)
        #self.printNode(self.rootptr,0)
        return ""
    def __repr__(self):
        return self.__str__()
    def printData(self,root):
        if root.left != None:
            self.printData(root.left)
        print root.data
        if root.right != None:
            self.printData(root.right)
        

    def printTree(self,root,level):
        if root.right != None:
            self.printTree(root.right,level+1)
        print " "*level*4+str(root) + "  "
        if root.left != None:
            self.printTree(root.left,level+1)



y = CityDB()
tmp = [[5000, (5067, 765)],
[1700, (1789, 7981)],
[1800, (1890, 8091)],
[2000, (2012, 212)],
[1900, (1901, 9101)],
[3000, (3012, 213)],
[8000, (8012, 218)],
[7500, (7567, 5767)],
[7600, (7678, 6877)],
[2500, (2567, 5762)],
[7800, (7890, 8097)],
[7700, (7789, 7987)],
[1500, (1567, 5761)],
[9000, (9012, 219)]]
for item in tmp:
    x = CityRecord(item[0], item[1][0],item[1][1])
    y.insert(x)
#print y
while 1:
    print y
    #print y.search(y.rootptr,int(raw_input("item1? ")),int(raw_input("item 2? ")))
    print y.remove(y.rootptr,int(raw_input("city code? ")))

File: Assignment 8/SuppliedFiles/CityDb.cpp
// FILE: CityDb.cpp
// CLASS IMPLEMENTED: city_db (see CityDb.h for documentation)
// INVARIANT of the ADT:
// ????
//
// Documentation for functions client don't have to know:
//
// void bst_remove_max(binary_tree_node<city_record>*& root_ptr,
//                     city_record& removed)
//   Precondition: root_ptr is a root pointer of a non-empty
//   binary search tree.
//   Postcondition: The city record with the largest city code
//   in the binary search tree has been removed, and root_ptr
//   now points to the root of the new (smaller) binary search
//   tree. The reference parameter, removed, has been set to a
//   copy of the removed city record.
//
// void bst_remove(binary_tree_node<city_record>*& root_ptr,
//                 const city_record& target)
//   Precondition: root_ptr is a root pointer of a binary search
//   tree (or may be 0 for the empty tree)
//   Postcondition: If target was in the tree, then one copy of
//   target has been removed, and root_ptr now points to the
//   root of the new (smaller) binary search tree. Otherwise,
//   if target was not in the tree, then the tree is unchanged.
//
// ...
//

// include relevant header files here
#include "CityDb.h"
#include "BinTree.h"
#include "CityRecord.h"
#include <iostream>

namespace cs3358Spring2007Assign08
{
   
   // NONMEMBER FUNCTIONS
   
   void bst_remove(binary_tree_node<city_record>*& root_ptr,
                 const city_record& target)
   {
      if (root_ptr->data() == target)
         delete bst_remove_max(root_ptr);// get rid of city code.
         return;
      bst_remove(root_ptr->left())
      bst_remove(root_ptr->right())
   }
   
   void bst_remove_max(binary_tree_node<city_record>*& root_ptr,
                     city_record& removed)
   {
      // we traverse the root_ptr -> left, taking all rights until we find the
      // largest value.
      if (root_ptr->left() != 0)
         binary_tree_node<city_record>* biggest = root_ptr->left();
         
         if (biggest->right() == 0) // nothing to the right of our node.
            biggest->set_left(root_ptr->left());
            biggest->set_right(root_ptr->right());
         while ( biggest->right() != 0)
            biggest = biggest -> right();
         
         removed = root_ptr->data();
         biggest->right()->set_left(root_ptr->left());
         biggest->right()->set_right(root_ptr->right());
         root_ptr = biggest->right();
         biggest->set_right(0);
         
            
      
      
      
      
   binary_tree_node<city_record>* searchRecur(int CityCode,
                                      binary_tree_node<city_record>* rootPtr)
   {
      if (rootPtr == 0) return 0;
      if (rootPtr->data().get_code() == CityCode)
         return rootPtr;
      binary_tree_node<city_record>* temp = searchRecur(CityCode,
                                                        rootPtr->left());
      if (temp != 0)
         return temp;
      temp = searchRecur(CityCode, rootPtr->right());
      if (temp != 0)
         return temp;
      return 0;
   }
   
   binary_tree_node<city_record>* searchRecur(int x, int y,
                                      binary_tree_node<city_record>* rootPtr)
   {
      if (rootPtr == 0) return 0;
      if (rootPtr->data().get_x() == x && rootPtr->data().get_y() == y)
         return rootPtr;
      binary_tree_node<city_record>* temp = searchRecur(x,y,
                                                        rootPtr->left());
      if (temp != 0)
         return temp;
      temp = searchRecur(x,y, rootPtr->right());
      if (temp != 0)
         return temp;
      return 0;
   }
   
   void insertRecur(city_record cr,
                                 binary_tree_node<city_record>* rootPtr)
   {
     assert(rootPtr != 0); //this should never happen.
     if (cr.get_code() <= rootPtr->data().get_code())
     {
          if (rootPtr->left() == 0)
          {
              rootPtr->set_left(new binary_tree_node<city_record>(cr));
              //rootPtr.left.data = cr;
          }
          else
              insertRecur(cr,rootPtr->left());
      }
      else
      {
          if (rootPtr->right() == 0)
          {
              rootPtr->set_right(new binary_tree_node<city_record>(cr));
          }
          else
              insertRecur(cr,rootPtr->right());
      }
   }
   
   
   
   
   
   
   
   
   
   
   void city_db::insert(city_record cr)
   {
      std::cout << "about to start operation."<<std::endl;
      if (root_ptr == 0) // no items, we can do this insert ourselves.
      {
         std::cout << "about to insert into the tree!!"<<std::endl;
         root_ptr = new binary_tree_node<city_record>(cr);
         std::cout << "inserted! phew!"<<std::endl;
         //root_ptr.data = cr;
      }
     else
         insertRecur(cr,root_ptr); 
   }
   
   void city_db::remove(int CityCode)
   {
   }
   void city_db::remove(int x, int y)
   {
   }
   bool city_db::search(int CityCode)
   {
      binary_tree_node<city_record>* temp = searchRecur(CityCode,root_ptr);
      if(temp != 0)
          return true;
      else
          return false;
   }
   
   bool city_db::search(int x, int y)
   {
      binary_tree_node<city_record>* temp = searchRecur(x,y,root_ptr);
      if(temp != 0)
          return true;
      else
          return false;
   }
      

   
    bool city_db::empty()
   {
      return (root_ptr == 0);
   }
   
   void city_db::printTree()
   {
      print(root_ptr, 0);
   }
   

}








